csp_billing_adapter/utils.py

```diff
@@ -73,4 +70,24 @@ def get_next_bill_time(
     elif billing_interval == 'test':
         kwargs['minutes'] = 5
 
-    return date + relativedelta(**kwargs)
+    return relativedelta(**kwargs)
+
+
+def get_next_bill_time(
+    date: datetime.datetime,
+    billing_interval: str
+) -> datetime.datetime:
+    """
+    Determine the next billing date using provided date and billing interval
+    """
+    return date + get_relative_period_delta(billing_interval)
+
+
+def get_prev_bill_time(
+    date: datetime.datetime,
+    billing_interval: str
+) -> datetime.datetime:
+    """
+    Determine the prev billing date using provided date and billing interval
+    """
+    return date - get_relative_period_delta(billing_interval)
```

Comment:
I am wondering if this should not be a config setting? Let me add he caveat that I could not find the section of the AWS doc that sets the limits on how far back we can bill.

What we are trying to address, if I remember this correctly, is that the adapter may not be able to bill for some undetermined amount of time and then, for whatever reason billing works again . But we are not allowed, this is the part I couldn't find in the doc anymore, to bill back to the beginning of the "api use problem". 

What I found was ```We recommend that you send a metering record every hour. However, you can aggregate usage over daily or monthly periods as well. If you experience an outage, you can aggregate buyer software use and send it in the following hours metering. You can't send more than one record per hour.``` in [1].

So if I extrapolate this to a month and let's say we have 3 months of outage we could actually report 3 hours in a row accumulating 1 month at a time and draining an "unpaid_records" cache one month at a time.

It may be worth to add a reference to the condition that prevents us from "bill back to the first access problem" as a comment.

Last but not least since this is csp specific, i.e. in some other CSP one may be allowed to "bill back to when the reporting issue started even if it is many months ago, it may be worth to add hooks to the csp implementations that modify this default behavior.


[1] https://docs.aws.amazon.com/marketplace/latest/userguide/container-metering-meterusage.html

Reply:
I do not remember where I read it or where I think I read it. This PR is to cover https://github.com/SUSE-Enceladus/csp-billing-adapter/issues/64. The issue you are addressing is https://github.com/SUSE-Enceladus/csp-billing-adapter/issues/65.

Reply:
Wrt to #65 if there is no rule on billing old usage and we are okay "truing up" in subsequent hours then nothing else needs to be done. This PR covers that because the last bill date is only updated once metering is successful. If that's all good but we want this configurable I think configurability needs to go into an issue for tracking purposes with target of v2.

Reply:
At present with this PR if the adapter starts up and has missed records it will true up because the records are determined by next bill time not by "now". And the next bill time is only incremented by one cycle at a time. This would look like:

- Startup and bill records for next bill time which will be in the past any number of cycles
- Every hour until "true up" is finished
  - Bill usage for the current next bill time
  - Update next bill time by one cycle
  
There's theoretically a chance that records in cache prior to next bill time at start of true up never get handled. But for that case we are way in weeds. And it can be covered in #65.

Reply:
Perhaps the routine's name needs to be changed if it is leading to confusion on this matter? It is being used to determine the start of the billing period associated with the cached `next_bill_time`...

However since we currently calculate each new `next_bill_time` value by adding the relative delta specified by the billing_interval to the previous `next_bill_time` value, the actual calculation being used in this function is appropriate.